File: Multiomics/TP/Clean/process.pl
#!/usr/bin/perl -w
use strict;
use warnings;
###################################################################################################
#思路：	对TP数据进行处理，将其处理成表达矩阵
		#1.读入文件夹内的所有TP文件
		#2.将其整理成TP矩阵
		#3.读入gencode.txt文件
		#4.比较并输出lncRNA与pc的表达矩阵,需要将ENSG转换为gene symbol
###################################################################################################

my @cancer_names=(
"BRCA",
"GBM",
"OV",
"LUAD",
"UCEC",
"KIRC",
"HNSC",
"LGG",
"THCA",
"LUSC",
"PRAD",
"SKCM",
"COAD",
"STAD",
"BLCA",
"LIHC",		#i=16
"CESC",
"KIRP",
"SARC",
"LAML",
"PAAD",
"ESCA",
"PCPG",
"READ",
"TGCT",
"THYM",
"KICH",
"ACC",
"MESO",
"UVM",
"DLBC",
"UCS",
"CHOL"
);


my %lnc_hash;
my %pc_hash;
#读入genecode中的lncRNA.txt数据
my $lnc_file="/mnt/5468d/gaomeihong/LncRNA3/GENCODE/Decompression/lncRNA.txt";
open LNC, $lnc_file or die "Can not open $lnc_file\n";
while(<LNC>)
{
	chomp;
	my @temp=split /\t/, $_ ;
	#print "$temp[0]\t$temp[1]\n";
	$lnc_hash{$temp[0]}=$temp[1];
}
close(LNC);
#读入genecode中的pc.txt数据
my $pc_file="/mnt/5468d/gaomeihong/LncRNA3/GENCODE/Decompression/pc.txt";
open PC, $pc_file or die "Can not open $pc_file\n";
while(<PC>)
{
	chomp;
	my @temp=split /\t/, $_ ;
	#print "$temp[0]\t$temp[1]\n";
	$pc_hash{$temp[0]}=$temp[1];
}
close(PC);


#构建哈希，存储文件名到样本名的映射


for my $i (0 .. $#cancer_names)
#for my $i (0 .. 0)
{
	print "$cancer_names[$i]\n";
	my $DIR_PATH="./Raw/TCGA-".$cancer_names[$i];
	my @filelist;
	opendir DIR, ${DIR_PATH} or die "Can not open $DIR_PATH\n";
	@filelist = readdir DIR;
	closedir(DIR);
	my $file;
	my $count=0;
	my @tp;
	my @lnc;
	my @pcg;
	foreach $file (@filelist)
	{
		my $tmep_index=index($file,".FPKM.txt");
		if(index($file,".FPKM.txt")>0)
		{
			#print "$file\n";
			open TP, "$DIR_PATH/$file" or die "Can not open $file\n";
			my $idx=0;
			my $lnc_idx=0;
			my $pcg_idx=0;
			while(<TP>)
			{
				chomp;
				my @temp=split /\t/, $_ ;
				if($count==0)
				{
					$tp[$idx][0]=$temp[0];
					$tp[$idx][1]=$temp[1];
					if(exists($lnc_hash{$temp[0]}))
					{
						#$lnc[$lnc_idx][0]=$temp[0];
						$lnc[$lnc_idx][0]=$lnc_hash{$temp[0]};
						
						$lnc[$lnc_idx][1]=$temp[1];
						$lnc_idx=$lnc_idx+1;
					}
					if(exists($pc_hash{$temp[0]}))
					{
						#$pcg[$pcg_idx][0]=$temp[0];
						$pcg[$pcg_idx][0]=$pc_hash{$temp[0]};
						
						$pcg[$pcg_idx][1]=$temp[1];
						$pcg_idx=$pcg_idx+1;
					}
				}
				else
				{
					$tp[$idx][$count+1]=$temp[1];
					if(exists($lnc_hash{$temp[0]}))
					{
						$lnc[$lnc_idx][$count+1]=$temp[1];
						$lnc_idx=$lnc_idx+1;
					}
					if(exists($pc_hash{$temp[0]}))
					{
						$pcg[$pcg_idx][$count+1]=$temp[1];
						$pcg_idx=$pcg_idx+1;
					}
				}
				$idx=$idx+1;
			}
			close TP;
			$count=$count+1;
		}
	}
	
	print "$#tp\n";
	print "$#lnc\n";
	print "$#pcg\n";
	my $res_tp_name="./New/".$cancer_names[$i].".tp.txt";
	open RES_TP, ">$res_tp_name" or die "Can not open $res_tp_name\n";
	for my $j (0 .. $#tp)
	{
		for my $k (0 .. $#{$tp[$j]})
		{
			print RES_TP "$tp[$j][$k]\t";
		}
		print RES_TP "\n";
	}
	close(RES_TP);
	
	my $res_lnc_name="./New/".$cancer_names[$i].".lncRNA.txt";
	open RES_LNC, ">$res_lnc_name" or die "Can not open $res_lnc_name\n";
	for my $j (0 .. $#lnc)
	{
		for my $k (0 .. $#{$lnc[$j]})
		{
			print RES_LNC "$lnc[$j][$k]\t";
		}
		print RES_LNC "\n";
	}
	close(RES_LNC);
	
	my $res_pcg_name="./New/".$cancer_names[$i].".pcg.txt";
	open RES_PCG, ">$res_pcg_name" or die "Can not open $res_pcg_name\n";
	for my $j (0 .. $#pcg)
	{
		for my $k (0 .. $#{$pcg[$j]})
		{
			print RES_PCG "$pcg[$j][$k]\t";
		}
		print RES_PCG "\n";
	}
	close(RES_PCG);
}

File: Multiomics/TP/Clean/process.R
###################################################################################################
#思路：	将表达数据处理为特征矩阵
		#1.读入lncRNA和pc文件
		#2.读入lnc和pcg的tp文件
		#3.将每行取均值
		#4.比较并输出lncRNA与pc的表达矩阵
###################################################################################################
cancer_names=c(	#33 cancers
"BRCA",
"GBM",
"OV",
"LUAD",
"UCEC",
"KIRC",
"HNSC",
"LGG",
"THCA",
"LUSC",
"PRAD",
"SKCM",
"COAD",
"STAD",
"BLCA",
"LIHC",		#i=16
"CESC",
"KIRP",
"SARC",
"LAML",
"PAAD",
"ESCA",
"PCPG",
"READ",
"TGCT",
"THYM",
"KICH",
"ACC",
"MESO",
"UVM",
"DLBC",
"UCS",
"CHOL"
)


TP_LNC_LIST=list()
TP_PCG_LIST=list()

TP_LNC_LIST[33+1]=NULL
TP_PCG_LIST[33+1]=NULL

names(TP_LNC_LIST)=cancer_names
names(TP_PCG_LIST)=cancer_names

lnc_path_name="/mnt/5468d/gaomeihong/LncRNA3/GENCODE/Decompression/lncRNA.txt"
pc_path_name="/mnt/5468d/gaomeihong/LncRNA3/GENCODE/Decompression/pc.txt"
LNC=read.table(file = lnc_path_name,
					sep = "\t", header = TRUE, 
					row.names = NULL, stringsAsFactors = FALSE)
PCG=read.table(file = pc_path_name,
					sep = "\t", header = TRUE, 
					row.names = NULL, stringsAsFactors = FALSE)

lp_path_name="/mnt/5468d/gaomeihong/LncRNA3/GAE_LGA/Datasets/Dataset3/lnc_pcg_net.txt"
LP=read.table(file = lp_path_name,
					sep = "\t", header = TRUE, 
					row.names = NULL, stringsAsFactors = FALSE)	

lnc_names=intersect(LP$Gene_name,LNC$gene_name)
pcg_names=intersect(colnames(LP),PCG$gene_name)

TP_LNC_FEAT=matrix(0,length(lnc_names),2*33+1)
TP_PCG_FEAT=matrix(0,length(pcg_names),2*33+1)
rownames(TP_LNC_FEAT)=lnc_names
rownames(TP_PCG_FEAT)=pcg_names	
TP_LNC_FEAT[,1]=lnc_names
TP_PCG_FEAT[,1]=pcg_names
				
feat_names=c("GeneSymbol")
for(i in 1:33)
{
	feat_name_temp=paste("TP_",cancer_names[i], sep = "", collapse = "")
	feat_name1=paste(feat_name_temp,"_MEAN", sep = "", collapse = "")
	feat_name2=paste(feat_name_temp,"_STD", sep = "", collapse = "")
	feat_names=c(feat_names,feat_name1,feat_name2)
}
colnames(TP_LNC_FEAT)=feat_names
colnames(TP_PCG_FEAT)=feat_names


#获取lnc_names和pcg_names的值
for(i in 1:length(cancer_names))
{
	print(i)
	tp_lnc_path_name=paste("./New/", cancer_names[i], ".lncRNA.txt", sep = "", collapse = "")
	tp_pcg_path_name=paste("./New/", cancer_names[i], ".pcg.txt", sep = "", collapse = "")
	
	TP_LNC=read.table(file = tp_lnc_path_name,
					sep = "\t", header = FALSE, 
					row.names = NULL, stringsAsFactors = FALSE)
	TP_PCG=read.table(file = tp_pcg_path_name,
					sep = "\t", header = FALSE, 
					row.names = NULL, stringsAsFactors = FALSE)
					
	TP_LNC<-TP_LNC[,-which(apply(TP_LNC,2,function(x) all(is.na(x))))]
	TP_PCG<-TP_PCG[,-which(apply(TP_PCG,2,function(x) all(is.na(x))))]
	
	
	TP_LNC=TP_LNC [which(TP_LNC[,1] %in% lnc_names),]
	TP_PCG=TP_PCG [which(TP_PCG[,1] %in% pcg_names),]
	
	TP_LNC=TP_LNC[!duplicated(TP_LNC[,1]),] #delete the duplicate lncRNA item
	TP_PCG=TP_PCG[!duplicated(TP_PCG[,1]),] #delete the duplicate gene item
	
	rownames(TP_LNC)=TP_LNC[,1]
	rownames(TP_PCG)=TP_PCG[,1]
	
	TP_LNC=TP_LNC[,-1]
	TP_PCG=TP_PCG[,-1]
	
	for(j in 1:length(lnc_names))
	{
		TP_LNC_FEAT[j,2*i]=mean(as.numeric(TP_LNC[j,]))
		TP_LNC_FEAT[j,2*i+1]=sd(as.numeric(TP_LNC[j,]))
	}


	for(j in 1:length(pcg_names))
	{
		TP_PCG_FEAT[j,2*i]=mean(as.numeric(TP_PCG[j,]))
		TP_PCG_FEAT[j,2*i+1]=sd(as.numeric(TP_PCG[j,]))
	}
	
	#未用到
	TP_LNC_LIST[[i]]=TP_LNC	
	TP_PCG_LIST[[i]]=TP_PCG
}

write.table(TP_LNC_FEAT, file ="TP_lnc_feature3.txt",sep="\t",row.names =FALSE, col.names =TRUE, quote =TRUE);
write.table(TP_PCG_FEAT, file ="TP_pcg_feature3.txt",sep="\t",row.names =FALSE, col.names =TRUE, quote =TRUE);	